drivers/spi/spi_wch.c

```diff
@@ -0,0 +1,252 @@
+}
+
+static int spi_wch_configure(const struct device *dev, const struct spi_config *config)
+{
+	const struct spi_wch_config *cfg = dev->config;
+	struct spi_wch_data *data = dev->data;
+	SPI_TypeDef *regs = cfg->regs;
+	int err;
+	uint32_t clock_rate;
+	clock_control_subsys_t clk_sys;
+	int8_t prescaler;
+
+	if (spi_context_configured(&data->ctx, config)) {
+		return 0;
+	}
+
+	if ((config->operation & SPI_HALF_DUPLEX) != 0U) {
+		LOG_ERR("Half-duplex not supported");
+		return -ENOTSUP;
+	}
+
+	if (SPI_OP_MODE_GET(config->operation) != SPI_OP_MODE_MASTER) {
+		LOG_ERR("Slave mode not supported");
+		return -ENOTSUP;
+	}
+
+	if ((config->operation & SPI_MODE_LOOP) != 0U) {
+		LOG_ERR("Loop mode not supported");
+		return -ENOTSUP;
+	}
+
+	if (SPI_WORD_SIZE_GET(config->operation) != 8) {
+		LOG_ERR("Frame size != 8 bits not supported");
+		return -ENOTSUP;
+	}
+
+	regs->CTLR1 = 0;
+	regs->CTLR2 = 0;
+	regs->STATR = 0;
+
+	if (spi_cs_is_gpio(config)) {
+		/* When using soft NSS, SSI must be set high */
+		regs->CTLR1 |= SPI_CTLR1_SSM | SPI_CTLR1_SSI;
+	} else {
+		regs->CTLR2 |= SPI_CTLR2_SSOE;
+	}
+
+	regs->CTLR1 |= SPI_CTLR1_MSTR;
+
+	if ((config->operation & SPI_TRANSFER_LSB) != 0U) {
+		regs->CTLR1 |= SPI_CTLR1_LSBFIRST;
+	}
+
+	if ((config->operation & SPI_MODE_CPOL) != 0U) {
+		regs->CTLR1 |= SPI_CTLR1_CPOL;
+	}
+
+	if ((config->operation & SPI_MODE_CPHA) != 0U) {
+		regs->CTLR1 |= SPI_CTLR1_CPHA;
+	}
+
+	clk_sys = (clock_control_subsys_t)(uintptr_t)cfg->clock_id;
+	err = clock_control_get_rate(cfg->clk_dev, clk_sys, &clock_rate);
+	if (err != 0) {
+		return err;
+	}
+
+	/* Approximate clock rate given ratios available */
+	prescaler = spi_wch_get_br(clock_rate / config->frequency);
+#if CONFIG_SPI_LOG_LEVEL >= LOG_LEVEL_INF
+	uint32_t j = 2;
+
+	for (int i = 0; i < prescaler; i++) {
+		j = j * 2;
+	}
+	LOG_INF("Selected divider %d, value %d, results in %d frequency", j, prescaler,
+		clock_rate / j);
+#endif
+	regs->CTLR1 |= prescaler << SPI_CTLR1_BR_POS;
+
+	data->ctx.config = config;
+
+	return 0;
+}
+
+static int spi_wch_transceive(const struct device *dev, const struct spi_config *config,
+			      const struct spi_buf_set *tx_bufs, const struct spi_buf_set *rx_bufs)
+{
+	const struct spi_wch_config *cfg = dev->config;
+	struct spi_wch_data *data = dev->data;
+	SPI_TypeDef *regs = cfg->regs;
+	int err;
+	uint8_t rx;
+
+	spi_context_lock(&data->ctx, false, NULL, NULL, config);
+
+	err = spi_wch_configure(dev, config);
+	if (err != 0) {
+		goto done;
+	}
+
+	spi_context_buffers_setup(&data->ctx, tx_bufs, rx_bufs, 1);
+
+	spi_context_cs_control(&data->ctx, true);
+
+	/* Start SPI *AFTER* setting CS */
+	regs->CTLR1 |= SPI_CTLR1_SPE;
+
+	while (spi_context_tx_on(&data->ctx) || spi_context_rx_on(&data->ctx)) {
+		if (spi_context_tx_buf_on(&data->ctx)) {
+			regs->DATAR = *(uint8_t *)(data->ctx.tx_buf);
+			while ((regs->STATR & SPI_STATR_TXE) == 0) {
+			}
+		} else {
+			regs->DATAR = 0;
+			while ((regs->STATR & SPI_STATR_TXE) == 0) {
```

Comment:
nittiest of nits: in other places this uses "== 0U" when testing if a bit is set or not. If you wish, change the file to use one way or the other.

My preference is 0.

Reply:
done